### DISCUSSION TITLE
Laravel Sanctum and Laravel Permissions

### DISCUSSION THREAD

**Discussion by AleJavierHY** [spatie/laravel-permission#2661] [Q&A]
I am using Laravel Permissions to create a Roles system in my web application that uses Laravel for the backend and Quasar with Vue on the front end. I have an api route that points to the index function of RoleController and it returns all the roles with their respective permissions and users related to each role.
Everything works correctly until I assign the auth:sanctum middleware to the api route. It is at that moment that the errors begin, specifically with the users relationship of the Laravel Permissions package. I already verified that it is only this relationship that does not work when I am using the auth:sanctum middleware in the api route of the index function.
The error it returns is the following:
`{
    "message": "Class name must be a valid object or a string",
    "exception": "Error",
    "file": "D:\\1-TRABAJO\\PROYECTOS\\ZUNspad2.0\\backend\\vendor\\laravel\\framework\\src\\Illuminate\\Database\\Eloquent\\Concerns\\HasRelationships.php",
    "line": 793,
}` Please, what could I be overlooking, I need to fix this error. The relationship exists and works but only when I do not use the auth:sanctum middleware

**Comment by parallels999** [comment]
[laravel-permission/v6/prerequisites#content-must-not-have-a-role-or-roles-property-nor-a-roles-method](https://spatie.be/docs/laravel-permission/v6/prerequisites#content-must-not-have-a-role-or-roles-property-nor-a-roles-method)

**Reply by AleJavierHY** [reply]
I extended the roles class because due to my business I saw the need to establish a column called parent_id through which a role will have a one-to-many relationship with another role.
Additionally, add two methods that establish such relationships. One is called parent and through it I get the parent role, the other relationship method is called children and returns the children of a role. The class extends Spatie's Role class and all relationships work fine until I use the auth:Sanctum Middleware. It is at that point that only the relationship with users fails.
The new model that extends Role is called Rol and its use is declared in the configuration file as the documentation says in these cases. When doing this in the controller I get the error I'm talking about: `Rol::query()->with("users")->get()`
It also fails when trying to get the permissions and their related users

**Reply by parallels999** [reply]
In my opinion there is some conflict with this package in your customizations, it is very difficult to know without being able to see the problem in a functional application

**Comment by drbyte** [comment]
It's difficult to know whether the problem is with your Controller, or with your custom model, or any other place.

Please post all the code necessary to recreate the problem on a new Laravel app.

**Comment by oommgg** [comment]
> I am using Laravel Permissions to create a Roles system in my web application that uses Laravel for the backend and Quasar with Vue on the front end. I have an api route that points to the index function of RoleController and it returns all the roles with their respective permissions and users related to each role. Everything works correctly until I assign the auth:sanctum middleware to the api route. It is at that moment that the errors begin, specifically with the users relationship of the Laravel Permissions package. I already verified that it is only this relationship that does not work when I am using the auth:sanctum middleware in the api route of the index function. The error it returns is the following: `{ "message": "Class name must be a valid object or a string", "exception": "Error", "file": "D:\\1-TRABAJO\\PROYECTOS\\ZUNspad2.0\\backend\\vendor\\laravel\\framework\\src\\Illuminate\\Database\\Eloquent\\Concerns\\HasRelationships.php", "line": 793, }` Please, what could I be overlooking, I need to fix this error. The relationship exists and works but only when I do not use the auth:sanctum middleware

I got the same issue too. and I add `sanctum` to `auth.php` => `guards`, then it's working 

```php
// config/auth.php

return [
    /** some other config ignored **/
    'guards'           => [
        'web' => [
            'driver'   => 'session',
            'provider' => 'users',
        ],

        'api' => [
            'driver'   => 'session',
            'provider' => 'users',
        ],

        'sanctum' => [
            'driver'   => 'sanctum',
            'provider' => 'users',  // <==  null for laravel-permission get the relations
        ],
    ],
    /** some other config ignored **/
];
```

now, I can use 

```php
return Role::withCount('users')->get();
```

without errors.

hope this helps you.
